File: find-all-duplicates-in-an-array/find-all-duplicates-in-an-array.py
class Solution:
    def findDuplicates(self, nums: List[int]) -> List[int]:
        
        size = len(nums)
        res = []
        for i in range(size):
            idx = abs(nums[i])-1
            
            if nums[idx]>0:
                nums[idx] = -nums[idx]
            else:
                res.append(idx+1)

        return res                

File: find-all-duplicates-in-an-array/find-all-duplicates-in-an-array.java
class Solution {
    public List<Integer> findDuplicates(int[] nums) {
        List<Integer> res = new ArrayList<>();
        
        int size = nums.length;
        int idx = 0;
        
        for(int i=0; i<size; i++){
            idx = Math.abs(nums[i])-1; // get idx of number          
            if(nums[idx]>0){ // check if index is already visited
                nums[idx] = -1 * nums[idx]; // mark the index as visited
            }
            else{
                
                res.add(Math.abs(nums[i]));
            }
            
        }
        
        return res;
    }
}